### PR TITLE
changes for freezed row move

### PR DESCRIPTION
**Pull request by tanvimehta10**
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Review by matthoward366 — COMMENTED** [review]
Should we be using `FROZEN` or `STATIC` instead of `FREEZED`?

**Comment by tanvimehta10** [comment]
> Should we be using `FROZEN` or `STATIC` instead of `FREEZED`?

Changed to static. Please verify

**Review by matthoward366 — COMMENTED** [review]
Haven't tested but the changes look good.  What happens if a user has the last row set as static but then deletes a row in the table?  Will the last row still be static?

**Comment by matthoward366** [comment]
@tanvimehta10 Take a look at https://github.com/elyra-ai/canvas/pull/735/checks?check_run_id=3262804031.  Looks like there are a number of linter errors.  You can run the build locally to see these.

**Comment by tanvimehta10** [comment]
> Haven't tested but the changes look good. What happens if a user has the last row set as static but then deletes a row in the table? Will the last row still be static?

Hi Matt, the static rows are defined using indexes. If the last row is deleted, the new last row will not be static.

**Comment by tanvimehta10** [comment]
@matthoward366 Fixed lint errors and verified by running the build locally
Screen Shot 2021-08-06 at 8 37 57 AM

File: canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx [hunk 1]
```diff
@@ -166,17 +195,32 @@ class MoveableTableRows extends React.Component {
 		const selected = this.props.controller.getSelectedRows(this.props.propertyId).sort(function(a, b) {
 			return a - b;
 		});
+		let leastValue = 0;
+		const staticRows = this.props.controller.getStaticRows(this.props.propertyId).sort();
+		if (staticRows && staticRows.length > 0 && staticRows.includes(0)) {
+			leastValue = staticRows[staticRows.length - 1];
+		}
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx:202]: I see this piece of code is the same as the one added in `topMoveRow()`. Would this make sense to be a separate function?

File: canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx [hunk 2]
```diff
@@ -215,7 +264,12 @@ class MoveableTableRows extends React.Component {
 			return a - b;
 		});
 		const controlValue = this.props.controller.getPropertyValue(this.props.propertyId);
-		for (var lastRow = selected[selected.length - 1]; lastRow < controlValue.length - 1; lastRow++) {
+		let maxValue = controlValue.length - 1;
+		const staticRows = this.props.controller.getStaticRows(this.props.propertyId).sort();
+		if (staticRows && staticRows.length > 0 && staticRows.includes(controlValue.length - 1)) {
+			maxValue = staticRows[0] - 1;
+		}
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx:271]: Same comment as above but for `downMoveRow()`

**Comment by tanvimehta10** [comment]
@caritaou The above code is more of assigning values to variable. Generally I would create a function where we have more complex logic to process the objects. Also these are very few lines to be moved to a function so that they can be reused. 
Not sure which approach is followed for Elyra. I could move it to a separate function if you suggest.

**Comment by tanvimehta10** [comment]
@caritaou Added a new function to handle the leastVal and maxVal. 
@matthoward366 Added this new feature to the side panel. You should be able to play around with the property values and test now.

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/structurelisteditor_paramDef.json [hunk 3]
```diff
@@ -10,6 +10,34 @@
         "World",
         "string",
         "Readonly phrase"
+      ],
```
Comment (caritaou) [inline canvas_modules/common-canvas/__tests__/test_resources/paramDefs/structurelisteditor_paramDef.json:13]: @tanvimehta10 Can you make these same changes to the corresponding structurelisteritor_paramDef in the test harness? We want the test files to be consistent to what we see in the test harness

Comment (tanvimehta10) [reply canvas_modules/common-canvas/__tests__/test_resources/paramDefs/structurelisteditor_paramDef.json:13]: Done

File: canvas_modules/harness/src/client/components/sidepanel-properties.jsx [hunk 4]
```diff
@@ -569,6 +601,43 @@ export default class SidePanelModal extends React.Component {
...
+					helperText='PropertyId format: {"name": "unique_id_for_control"}'
+				/>
+			</div>
+		);
+
+		const setStaticRowsIndexes = (
+			<div className="harness-sidepanel-children" id="sidepanel-properties-set-static-rows-indexes">
+				<TextInput
+					labelText="Set the indexes for static rows"
+					id="harness-propertyId-staticRows"
```
Comment (caritaou) [inline canvas_modules/harness/src/client/components/sidepanel-properties.jsx:622]: Can you give these unique ids?
image

File: canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx [hunk 5]
```diff
@@ -141,19 +158,49 @@ class MoveableTableRows extends React.Component {
...
+		const staticRows = this.props.controller.getStaticRows(this.props.propertyId).sort();
+		for (var firstRow = selected[0]; firstRow > leastValue; firstRow--) {
 			for (var i = 0; i <= selected.length - 1; i++) {
-				const selectedRow = selected.shift();
-				const tmpRow = controlValue[selectedRow - 1];
-				controlValue[selectedRow - 1] = controlValue[selectedRow];
-				controlValue[selectedRow] = tmpRow;
-				selected.push(selectedRow - 1);
-				this.props.controller.moveErrorMessageRows(this.props.propertyId.name, selectedRow, selectedRow - 1);
+				if (staticRows.length > 0 && selected[0] > leastValue + 1) {
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx:189]: These if cases have the same code in the block. Can you combine them into a single if statement?

File: canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx [hunk 6]
```diff
@@ -166,17 +213,29 @@ class MoveableTableRows extends React.Component {
...
-				if (selectedRow !== 0) {
+				if (staticRows.length > 0 && selected[0] > leastValue + 1) {
+					const selectedRow = selected.shift();
 					const tmpRow = controlValue[selectedRow - 1];
 					controlValue[selectedRow - 1] = controlValue[selectedRow];
 					controlValue[selectedRow] = tmpRow;
 					selected.push(selectedRow - 1);
 					this.props.controller.moveErrorMessageRows(this.props.propertyId.name, selectedRow, selectedRow - 1);
+				} else if (staticRows.length === 0) {
+					const selectedRow = selected.shift();
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx:231]: Are you able to combine these if statements as well? Is there a reason why this line `const selectedRow = selected.shift();` is outside the if statement?

File: canvas_modules/common-canvas/src/common-properties/properties-controller.js [hunk 7]
```diff
@@ -1746,4 +1746,57 @@ export default class PropertiesController {
...
 		return this.propertiesStore.getAddRemoveRows(propertyId);
 	}
+
+	/**
+	 * Freeze row move buttons for row indexes in given array
+	 * @param propertyId The unique property identifier
+	 * @param rowIndexes Array of row indexes
+	 *
+	 */
+
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/properties-controller.js:1756]: Can you remove this extra line?

File: canvas_modules/common-canvas/src/common-properties/properties-controller.js [hunk 8]
```diff
@@ -1746,4 +1746,57 @@ export default class PropertiesController {
...
+		}
+	}
+
+	clearStaticRows(inPropertyId) {
+		const propertyId = this.convertPropertyId(inPropertyId);
+		this.propertiesStore.clearStaticRows(propertyId);
+	}
+
+	/**
+	 * Validate if the array for freeze rows is correct. Should only have continuos value of row indexes
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/properties-controller.js:1778]: Typo for `continuos`

File: canvas_modules/common-canvas/src/common-properties/properties-controller.js [hunk 9]
```diff
@@ -1746,4 +1746,57 @@ export default class PropertiesController {
...
+	 * Must not contain first and last row index together in the array ever. you can only freeze either first row or the last row
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/properties-controller.js:1779]: > you can only freeze either first row or the last row
This should be "You can only freeze either first n rows or the last n rows"?

File: canvas_modules/common-canvas/src/common-properties/reducers/row-static.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2017-2020 Elyra Authors
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/reducers/row-static.js:2]: Year is 2021

File: canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx (hunk 6 above)
Comment (tanvimehta10) [reply canvas_modules/common-canvas/src/common-properties/components/moveable-table-rows/moveable-table-rows.jsx:231]: We need this condition for row move up. If I remove this the behavior is incorrect.

**Comment by tanvimehta10** [comment]
@caritaou Added all the changes requested by you and also fixed lint errors from the build failure

**Review by caritaou — APPROVED** [review]

**Review by matthoward366 — APPROVED** [review]
lgtm

**Pull request merged**